### PR TITLE
Add filters.hag_dem

### PR DESCRIPTION
**Pull request by jtfell**
Created a new filter for calculating HAG values from a raster-based DEM based on feedback from  #2881

### CODE REVIEW

**Comment by chambbj** [comment]
This looks very nice @jtfell. I personally would only suggest updating the copyright for current year and your own name/contact info as this is now unrelated to the original HAG filter in that regard.

**Comment by jtfell** [comment]
Cheers @chambbj - I've updated the copyright info

File: filters/HAGDEMFilter.cpp [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+
+HAGDEMFilter::HAGDEMFilter()
+{}
+
+
+void HAGDEMFilter::addArgs(ProgramArgs& args)
+{
+    args.add("raster", "GDAL-readable raster to use for DEM (uses band 1, "
```
Comment (abellgithub) [inline filters/HAGDEMFilter.cpp]: If you use `args.add(...).setPositional()`, you will this option will be required and you won't need to check that it was set in `prepared()`.

File: filters/HAGDEMFilter.cpp [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    m_raster.reset(new gdal::Raster(m_rasterName));
+    m_raster->open();
+}
+
+void HAGDEMFilter::prepared(PointTableRef table)
+{
+  if (m_rasterName == "")
+    throwError("Must specify a 'raster' DEM");
+  if (m_band <= 0)
+    throwError("'band' must be greater than 1");
```
Comment (abellgithub) [inline filters/HAGDEMFilter.cpp]: greater than 0

File: filters/HAGDEMFilter.cpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    // If "zero_ground" option is set, all ground points get HAG of 0
+    if (m_zeroGround && point.getFieldAs<uint8_t>(Id::Classification) == ClassLabel::Ground)
+    {
+        point.setField(Id::HeightAboveGround, 0);
+    }
+    else
+    {
+        double x = point.getFieldAs<double>(Id::X);
+        double y = point.getFieldAs<double>(Id::Y);
+        double z = point.getFieldAs<double>(Id::Z);
```
Comment (abellgithub) [inline filters/HAGDEMFilter.cpp]: The fetch of Z could be deferred until you know you need it.

File: filters/HAGDEMFilter.cpp (hunk 2 above)
Comment (jtfell) [reply filters/HAGDEMFilter.cpp]: I was following the convention from https://github.com/PDAL/PDAL/blob/master/filters/DEMFilter.cpp#L104 where the band counts from 1. Happy to change it, but I thought it best to be consistent.

Comment (hobu) [reply filters/HAGDEMFilter.cpp]: Bands in GDAL count from one. A band of 0 is an error. The conditional is bad (in both places!) and should be changed to `m_band < 0`.

Comment (abellgithub) [reply filters/HAGDEMFilter.cpp]: The text needs to say "must be greater than 0"

Comment (jtfell) [reply filters/HAGDEMFilter.cpp]: I agree that the error message is wrong but it seems to me that the conditional is fine.

Comment (jtfell) [reply filters/HAGDEMFilter.cpp]: I have pushed a fix for the error message (in both places) @abellgithub @hobu - Does this look right to you?

**Comment by abellgithub** [comment]
Merged into https://github.com/PDAL/PDAL/tree/issue-2886
Removed static type on vector (bad in case people want to thread) and made formatting changes.

**Comment by abellgithub** [comment]
@jtfell : Do you have a test for this that I can integrate?

**Pull request closed without merging**